### DISCUSSION TITLE
Upgrading

### DISCUSSION THREAD

**Discussion by G0nz0uk** [grafana/grafana-csv-datasource#31] [Q&A]
Hello,

Apologies as it’s not an issue, but wasn’t sure how to contact you.

What a great plugin, I’ve been using it for a while on my Linux Grafana boxer work.

How do I check what version I’m using?
How can I upgrade to the latest?

thanks

**Comment by marcusolsson** [comment]
Thanks, and no worries!

I've enabled [Discussions](https://github.com/marcusolsson/grafana-csv-datasource/discussions) if you have future questions. It might be useful for anyone with the same question as issues tend to disappear after they're closed.

You should be able to see the current version if you run:

```
grafana-cli plugins ls
```

To update you should be able to run:

```
grafana-cli plugins update marcusolsson-csv-datasource
```

Hope that helps!

**Comment by G0nz0uk** [comment]
Thanks,  I’m now on version 0.3.0 but all my CSV dashboards no longer work.

Can I put 0.1.0 back on easily?

I have a snapshot I can roll back to also.

**Comment by G0nz0uk** [comment]
I have rolled back the snapshot and back on version 0.1.0 and it’s now working again.

**Comment by marcusolsson** [comment]
That's a shame :( I'd love to know more about the issues you're seeing so that you can update. Are you getting any errors?

**Comment by G0nz0uk** [comment]
I don’t see any errors, just my panels don’t display anything.  I have around 10 csv data sources I use for Citrix.  They all stopped when I went to 0.3.0, so rolled back.

I just point to the CSV file on an IIS server that I produce via powershell.  In Grafana I just use the Delimiter as Comma and it shows.

How do I get to the error logs?

Is there a method I can upgrade to 0.3.0 again and roll back to 0.1.0 again?

Thanks

**Comment by andywhiteglobal** [comment]
Is there a quick way to downgrade from 0.3.0 to 0.1.0 so I can look at what the differences might be?

**Comment by marcusolsson** [comment]
This should work:

```
grafana-cli plugins install marcusolsson-csv-datasource 0.1.0
```

It would be really helpful for me know whether this is an issue with connecting to the CSV file, or whether it's related to the actual CSV content.

For example, does it work if you [create a gist](https://gist.github.com/) with the same CSV content and add it as a data source?

**Comment by G0nz0uk** [comment]
I'll try the downgrade and gist test.

I tried updating again and the error I get is "unsupported storage type"

**Comment by marcusolsson** [comment]
Oh! I'm sorry about that. v0.3.0 introduced support for local files. In your data source config, you can now select between two different storage types: HTTP and Local.

The plugin should've defaulted to HTTP for existing users, but it seems like it didn't. Try going to the data source config page and clicking 'HTTP' (that should set the storage type). If that doesn't work (and you don't want to recreate the data source) we can edit the dashboard JSON instead to add it.

**Comment by G0nz0uk** [comment]
I have changed it from local to http and when I hit test I get:

Get "https://myiisserver.local/myoutput.csv": net/http: invalid header field name ""

I have 8 CSV datasources and they all seem to do the same.  These are powershell scripts that output to CSV.

If I downgrade to 0.1.0 all is good again.

I use the Delimeter as Tab

**Comment by marcusolsson** [comment]
Hm. I wonder if this is related to Custom headers. It looks like the plugin is sending an empty HTTP header. Are you currently setting any custom HTTP headers?

**Comment by G0nz0uk** [comment]
Hmm sorry could you explain that a little more in terms of custom headers?  

One of the csv's looks like this, I've had to remove private information though.

<img width="1006" alt="Screenshot at Jan 19 16-54-55" src="https://user-images.githubusercontent.com/15994497/105066843-18e1ee80-5a77-11eb-95a3-12874ae837c8.png">

If I look on IIS which doesn't have excel and opens with Wordpad it looks like this for the first 2 lines:

"Users","Apps","LogonTime","UserIP","UserDevice","VMIP","VDI Name","DG","SessionState","PowerState","FaultState"
"DOMAIN\joe","Word","19/01/2021 09:03:34","10.1.2.3","MyPC","10.5.6.7","VDIW10123","W10_Standard","Active","On","None"

**Comment by marcusolsson** [comment]
I think I've managed to reproduce this. Will do some troubleshooting on my part and update you as soon as I know more.

**Comment by G0nz0uk** [comment]
wow thanks!

**Comment by marcusolsson** [comment]
I've published 0.3.1, which contains a fix, arguably for the symptom rather than the cause. Could you try and see if that fixes your problem?

**Comment by G0nz0uk** [comment]
I get the same issue I'm afraid on all the dashboards we run.  Just says 'unsupported storage' type.

If I then go and look at the data source and test it I get the all green, but of I go back to the panel in no longer says 'unsupported storage' but 'parse error on line 1, column 22: bare " in non-quoted-field'

**Comment by G0nz0uk** [comment]
yes it's using 
5.4.0-62-generic #70-Ubuntu SMP Tue Jan 12 12:45:47 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux

**Comment by G0nz0uk** [comment]
Hi @marcusolsson I was just wondering if there is anymore testing I can do?

**Comment by marcusolsson** [comment]
I haven't had a chance to look at this yet, unfortunately.

Do you see the `'parse error on line 1, column 22: bare " in non-quoted-field'` for all your data sources, or is it only certain CSV files?

Also, in the meantime, could you maybe try removing the data source and adding it back again to make sure there's no settings from the previous versions that interfere?

I'll try to get to this during the week. I haven't heard about this issue from other users, so I'm thinking it's related to configuration or your specific data. 

Thanks for your patience on this!

**Comment by G0nz0uk** [comment]
So I have just updated to 0.3.1 and restarted Grafana and all my CSV panels how as 'Unsupported file type'.

I added a new datasource and tested it and it said 'ok' I then when and created a new cvs panel table and got the 'parse error on line 1, column 22: bare " in non-quoted-field'

Rolled back to 0.1.0 and all 6 panels are working again.

**Comment by marcusolsson** [comment]
I googled on the error and it looks like it's due to a single double-quote in the cells.

For example,

```
foo, bar, tes"t, baz
```

or due to single quote in a quoted cell

```
"I", "can't", "even"
```

It looks like we can allow this by enabling lazy quotes:

From https://golang.org/pkg/encoding/csv/:

```go
    // If LazyQuotes is true, a quote may appear in an unquoted field and a
    // non-doubled quote may appear in a quoted field.
    LazyQuotes bool
```

Not sure why this is working in 0.1.0 though. Is there any way you could verify if this is indeed what causes the error?

**Comment by marcusolsson** [comment]
I've published 0.3.2 that enable lazy quotes. Let me know if it works.

**Comment by G0nz0uk** [comment]
Thanks, I will test in the morning and let you know how I get on.  It it works I’ll see where these lazy quotes are, some of the CSVs have people names in them which have apostrophes in them, could that be it?

**Comment by G0nz0uk** [comment]
Same issue, this is the output of one of the csv files.  I've removed the other users as it would be long.  It has the column names at the top and then the user session info for their Citrix session:

"Users","Apps","LogonTime","UserIP","UserDevice","VMIP","VDI Name","DG","SessionState","PowerState","FaultState"
"DOMAIN\bob","CSS Live LD_1-1","04/02/2021 15:58:53","10.17.217.244","HHNWL414","10.217.14.7","LD-W1042","W10_Standard","Active","On","None"

what I see

![image](https://user-images.githubusercontent.com/15994497/106931618-acb9e880-670e-11eb-9623-2a521e118bbd.png)

**Comment by marcusolsson** [comment]
Thanks for the test data. It seems to work for me on Grafana 7.3.6 with the latest plugin version 0.9.2.

<img width="868" alt="Screen Shot 2021-02-05 at 07 02 30" src="https://user-images.githubusercontent.com/8396880/106995886-2938df80-6780-11eb-8846-4b0041d081ff.png">

**Comment by G0nz0uk** [comment]
Can I send you the whole csv to test?

I'm on 7.3.7, how do I get my plugin version as I though you were on 0.3.2 not 0.9.2?

https://pastebin.com/CKRkQrb0

![image](https://user-images.githubusercontent.com/15994497/107027785-88105000-67a4-11eb-8587-d3fe4578e8c0.png)


Thanks

**Comment by marcusolsson** [comment]
Sorry, I'm mixing up the versions from my plugins 😅 0.3.2 is indeed the latest one.

I'm able to load the CSV you linked to successfully. I tried loading it one 0.2.0 and 0.3.0 which also worked. It leads me to believe that the error is related to your data source config rather than the data itself. 

Are you still getting the `bare " in non-quoted-field` error?

(Also, feel free to delete the CSV data on the link you sent me)

**Comment by G0nz0uk** [comment]
I got the 'Unsupported file type'

Do you have a screenshot of you datasource so I can compare?

my csv's sit on IIS btw, I guess I could try locally to see if it's that?

**Comment by marcusolsson** [comment]
Do you mean 'Unsupported storage type', or _file_ type? I thought you got past that one?

> but of I go back to the panel in no longer says 'unsupported storage'

This is what I did:

1. I copied the link to the raw CSV that you linked in this thread
2. Created a new CSV data source
3. Pasted the URL
4. Went to Explore—works without any changes to query editor

**Comment by G0nz0uk** [comment]
1. I stop grafana
2. update from 0.1.0 to 0.3.2
3. I start Grafana
4. Check my CSV panels
![image](https://user-images.githubusercontent.com/15994497/107077073-443d3b00-67e4-11eb-80e4-bdca40e2449a.png)

csv's just live in IIS on a remote server

Scratching my head.  I have Grafana at home so might try there, but don't have a web server.

**Comment by marcusolsson** [comment]
Ok, I've published 0.3.3 that defaults to HTTP if there's no storage type set. Could you try that one maybe?

**Comment by G0nz0uk** [comment]
Sorry about this issue as it's only me, I will try on my home lab too and see what the results are.  I have just upgraded to 0.3.2 ( whoops I mean 0.3.3) and all my panels now say:

![image](https://user-images.githubusercontent.com/15994497/107112779-1bec2580-6852-11eb-8885-679bdf7a100f.png)

Roll back to 0.1.0 and all is good.

I wonder if it's the upgrade process?  I wonder if you created a 0.1.1 (or something you haven't used) and made no changes to 0.1.0 and then I try the upgrade?  And see if it still works.

**Comment by marcusolsson** [comment]
No worries. We'll get it working. Did you try 0.3.3 yet?

**Comment by G0nz0uk** [comment]
Sorry that was a typo, I meant 0.3.3.

I’ve corrected my previous post.

**Comment by marcusolsson** [comment]
That feels like progress though. I _think_ that error tells you that you have rows with different number of columns in your data.

**Comment by G0nz0uk** [comment]
Update:

So I set this all up at home on my 7.3.7 Grafana server with CSV 0.1.0 and the csv is good like at work.

I upgrade to 0.3.3 and I get:

![image](https://user-images.githubusercontent.com/15994497/107123595-48c22c00-6896-11eb-86ce-f1f15aea3cfa.png)

Roll back to 0.1.0 and all is good again.

**Comment by marcusolsson** [comment]
Are you using the same data as before? I can reproduce your error with the following CSV:

```
foo, value
john, 10, foo
lily, 20
```

Notice that the second line has three columns, whereas the others have two. Same thing happens if you have too few columns:

```
foo, value
john
lily, 20
```

Could you look at line 8 in your CSV data and see if it's missing, or has more columns than there are headers?

**Comment by G0nz0uk** [comment]
Yeah I'm using https://pastebin.com/Y9HcXr87

I just saved as a csv in Excel and put in on my IIS server.

![image](https://user-images.githubusercontent.com/15994497/107125722-6944b300-68a3-11eb-81d1-8f332ed62375.png)

and when I role back to 0.1.0

![image](https://user-images.githubusercontent.com/15994497/107125791-c2144b80-68a3-11eb-86ac-bd095c54dd1b.png)

Could it be the empty cell in line 7 rather that line 8?

**Comment by marcusolsson** [comment]
Hm. I wonder if this is a Windows-specific thing, since I have no trouble reading the file you just gave me (macOS). 

Is there any way you can remove the offending lines until you get it working, and then send me the specific lines?

**Reply by G0nz0uk** [reply]
Yeah I’ll try that and keep you posted and also try on the different Grafana’s I have.  I use one in Docker Compose so I might run a slightly older version and see too.

I’m not sure it’s Windows or a MacOS thing as I’m tried that on both as my work is Mac and home is Windows, I’ve even tried different browsers.

**Comment by G0nz0uk** [comment]
Hi, right, I've built a fresh Grafana server this morning on 7.3.7.

I've started with version 0.1.0 and this csv I made in Excel

![image](https://user-images.githubusercontent.com/15994497/107140781-57532680-691c-11eb-9449-a852c06132bc.png)

Created the datasource.

![image](https://user-images.githubusercontent.com/15994497/107140810-8cf80f80-691c-11eb-9121-f78d05378b94.png)

Created a panel.

![image](https://user-images.githubusercontent.com/15994497/107140829-c92b7000-691c-11eb-80ab-8019f998aa03.png)

upgraded to 0.3.3

![image](https://user-images.githubusercontent.com/15994497/107140850-fd9f2c00-691c-11eb-81a0-9ab12d57c506.png)

Then re ran the query

![image](https://user-images.githubusercontent.com/15994497/107140876-1d365480-691d-11eb-815b-837bbddd75a0.png)

How strange eh?

Also on this version I can make up a datasource that doesn't exist and it accepts it, this csv doesn't exist.

![image](https://user-images.githubusercontent.com/15994497/107141018-047a6e80-691e-11eb-96ed-520ba945da11.png)


If I roll back to 0.1.0

![image](https://user-images.githubusercontent.com/15994497/107140935-7900dd80-691d-11eb-82bd-5a03a0dfc9dd.png)

and test the datasource

![image](https://user-images.githubusercontent.com/15994497/107141451-b0bd5480-6920-11eb-81f4-15c0adc612f0.png)


Then correct it

![image](https://user-images.githubusercontent.com/15994497/107141503-efeba580-6920-11eb-8682-23f95c9004e3.png)


run the query

![image](https://user-images.githubusercontent.com/15994497/107140979-c715e100-691d-11eb-87ea-0b609bc7082e.png)

Weird isn't it?

**Comment by marcusolsson** [comment]
Weird indeed. I'm wondering whether it makes a difference if you generate files from Excel. Line breaks and file ending etc (yes, wild guesses at this point). Just so we can make sure that we're using the exact same data, could you create a data source pointing at this Gist I made?:

```
https://gist.githubusercontent.com/marcusolsson/f2c392e4bb0806affc80313498c8f3a6/raw/700dc394c4a7cad0bdcffe6306b8e73b8b331660/gonzo.csv
```

**Reply by G0nz0uk** [reply]
So before on 0.1.0 you see mine and yours:

![image](https://user-images.githubusercontent.com/15994497/107142395-8bcbe000-6926-11eb-9758-abf26bd9e5ea.png)

Upgrade to 0.3.0

Yous is ok and mine has issue son line 8

![image](https://user-images.githubusercontent.com/15994497/107142416-b027bc80-6926-11eb-8417-4cc64dbe6aad.png)

![image](https://user-images.githubusercontent.com/15994497/107142425-be75d880-6926-11eb-9b28-8f411ffaf3a1.png)

**Reply by G0nz0uk** [reply]
If I put the same csv in gisthub it works, so is it IIS, Excel?

![image](https://user-images.githubusercontent.com/15994497/107143552-9e95e300-692d-11eb-9bf9-c12499686738.png)

https://gist.githubusercontent.com/G0nz0uk/877c56160fe4b760e499535dbff9ee49/raw/27d733a0ee1efedacecebc394c58b956b68432c4/gonzo.csv

**Comment by G0nz0uk** [comment]
Also just upgraded works 7.3.7 Grafan to 7.4.0 and all the CSV data sources went grey, upgraded to 0.3.3 and all the data sources have been found and but get the issue on line 8 again.

Think is I can't upgrade to 7.4.0 with 0.1.0 now, not sure if you are aware of that?

**Reply by marcusolsson** [reply]
I'm not sure I'm aware. Are you getting errors when installing 0.1.0 on 7.4?

**Reply by G0nz0uk** [reply]
So I'm on 0.1.0 with all my nice CSV dashboards working with Grafana 7.3.7.

I upgrade to 7.4

![image](https://user-images.githubusercontent.com/15994497/107228350-7b922e80-6a14-11eb-988b-e257fbeb8e50.png)

All my CSV datasources go grey, all other plugins are fine.

![image](https://user-images.githubusercontent.com/15994497/107228666-d2980380-6a14-11eb-8519-11e1d568f971.png)

I then upgrade to 0.3.3 and they are ok again

![image](https://user-images.githubusercontent.com/15994497/107228793-f8250d00-6a14-11eb-8993-998eb36d161a.png)

but we get the usual

![image](https://user-images.githubusercontent.com/15994497/107228897-15f27200-6a15-11eb-8615-2a5ae98eb157.png)

These csv's are creating by powershell and dumped to an IIS server

Roll back to 0.1.0 and 7.3.7 and all is good again.

However your gist url works when on 7.4 and 0.3.3 https://gist.githubusercontent.com/marcusolsson/f2c392e4bb0806affc80313498c8f3a6/raw/700dc394c4a7cad0bdcffe6306b8e73b8b331660/gonzo.csv

I wonder if the upgrade process causes issues.  Are you able to clone 0.1.0 to another number and publish and I'll upgrade to that, in theory it should work as nothing has changed?

Thanks

**Reply by marcusolsson** [reply]
Could you check in the list of Plugins to see if the CSV plugin is being loaded properly? It seems like there's a bug in 7.4 that incorrectly reports plugins as having a modified signature (so far it only seems to affect Windows users). I'll keep [this issue](https://github.com/marcusolsson/grafana-csv-datasource/issues/32) updated.

Hm. I'm thinking how we could pinpoint the error. Could you diff the responses from your IIS and a GitHub Gist to see if there's any weird encoding happening?

**Reply by G0nz0uk** [reply]
With 7.4.0 and 0.1.0 I get this:

![image](https://user-images.githubusercontent.com/15994497/107258465-bd7e9d00-6a33-11eb-80a7-612f7aa08a21.png)

with 0.3.3 which I can't use I get

![image](https://user-images.githubusercontent.com/15994497/107258550-d9823e80-6a33-11eb-99c9-37abb6e6353f.png)

**Reply by G0nz0uk** [reply]
Hi, is there a way to use 0.1.0 in 7.4?

Could you explain how I can diff the responses from IIS and a GitHub Gist?

**Reply by marcusolsson** [reply]
Unfortunately, there's a bug in Grafana 7.4.0 that blocks signed plugins from being loaded properly. I think you can load it if you allow unsigned plugins in your grafana.ini.

As for diffing the responses, I'm not sure how you would do it in Windows, but I would've done something like this on a Unix machine:

```
curl https://gist.githubusercontent.com/G0nz0uk/877c56160fe4b760e499535dbff9ee49/raw/27d733a0ee1efedacecebc394c58b956b68432c4/gonzo.csv > gist.csv
curl https://iis.server.com/data.csv > iis.csv
diff gist.csv iis.csv
```

**Reply by G0nz0uk** [reply]
I wonder if it worth staying on 7.3.7 until they fix it. 
 
I did see in Grafana.ini there is:
 
`;allow_loading_unsigned_plugins =`
 
But I’m not sure what to type for the plugin name after the = symbol, plus I need to manually download 0.1.0 (do you have a link?) and put in a Grafana folder for which I’m not sure.
 
I’ll look into the diff commands.

**Comment by marcusolsson** [comment]
You can download any plugin version from the [Releases](https://github.com/marcusolsson/grafana-csv-datasource/releases) (under Assets).

`allow_loading_unsigned_plugins` takes a comma-separated list of plugin ids. For this plugin, you should put `marcusolsson-csv-datasource` as the value.

**Reply by G0nz0uk** [reply]
Thanks I'll try that.

Right I've done those curl tests.  What I learn't is on Windows it's 'cUrl' and works the same.

Now here is the strange bit, maybe I'm missing something.

Grafana is on 7.3.0 and 0.1.0

As you know CSVs work.

From a windows server I copied one of the working csv's using the datasource path (using example names):
cUrl https://iisserver/file.csv > iis.csv

I then copied the above to gist and created a new datasource and the table shows just great.

So I then export it from the Grafana server:
curl https://gist.githubusercontent.com/G0nz0uk/blahblahblah/gist.csv > gist.csv

I then copied the 2 files to a Windows server and opened Notepad ++ and the compare tool and it said both files are the same.

I then upgraded to 0.3.3 and get the usual issue on line 8.

When I do a 'diff gist.csv iis.csv' as I copied them to grafana then it's harder to see but they look the same, but that tool is more of a manual check.

**Reply by G0nz0uk** [reply]
Sorry for all these updates:

So I've just stayed on 7.3.7 and upgraded to 0.3.3 and broke the csv's as usual that point to the remote IIS server (we know gist works).

I then decided to go to the IIS server and copy one of the csv's that no longer works and copy it locally to the Grafana server (Ubuntu) and create a new datasource but use the local option and the csv works.

So on 7.3.7 with 0.3.3 my csv's copied locally work but not when on the remote IIS server.  I tried at home and the same thing happens as I thought maybe it's the IIS server at work.

My head hurts.

**Comment by G0nz0uk** [comment]
Could it be IIS based on above?

Can you think of another quick and easy Web server I can run to test this on?  This might prove if it's IIS as I've tried it on 2 IIS servers now.  Local is fine and gist.

**Reply by marcusolsson** [reply]
If you have Python installed, you could try running:

```
python -m http.server 8000
```

in the same directory as the CSV file. You should be able to access it through `http://localhost:8000/file.csv`.

**Reply by G0nz0uk** [reply]
I think it's all Windows IIS web servers.

Tonight I tested at home on my Grafana 7.3.7 with CSV 0.3.3 all using the same csv file.

Datasources and results:

Gist Link - Worked
Local CSV - Worked
Python Website on a new Ubuntu VM- Worked
IIS on Windows 2016 Server - **Failed with row 8 error**
IIS on my Win 10 laptop - **Failed with** 

![image](https://user-images.githubusercontent.com/15994497/107997634-1a052d80-6fdb-11eb-8f00-d1fded250d0b.png)

When I rolled back to 0.1.0 my Wn10 IIS server was ok

![image](https://user-images.githubusercontent.com/15994497/107997957-f1316800-6fdb-11eb-930b-7142ba82e5ab.png)

Rolled back to CSV 0.1.0 all worked. 

All work I did the same tests and on 0.3.0 all the datasources pointing to IIS failed, gist, local and python websites worked.

Rolled back to CSV 0.1.0 and all worked.

**Reply by marcusolsson** [reply]
Oh. I might have an idea of what could cause this. By default, the plugin sets `Accept: text/csv` on the request, and it looks like IIS is configured to block requests for CSV content.

If you go to the data source config and add a Custom HTTP Header for:

**Header:** Accept
**Value:** text/plain

This will override the default `text/csv` value. Let me know if that works!

**Reply by G0nz0uk** [reply]
like this?

![image](https://user-images.githubusercontent.com/15994497/108053670-a1878680-7045-11eb-9413-5248be5ec549.png)

**Reply by marcusolsson** [reply]
Assuming you wrote `text/plain` as value, yes :)

**Reply by G0nz0uk** [reply]
Unfortunately the same errors on all IIS based panels.  I've updated all the custom headers to use what you put above too.

**Reply by G0nz0uk** [reply]
I wonder if anyone else can test with IIS on 0.1.0 and 0.3.3

**Reply by marcusolsson** [reply]
Maybe now that we've know more, could you create an issue for this and we can continue the discussion over there?

**Reply by G0nz0uk** [reply]
I scripted all CSVs to go from our Windows servers to the Grafana Ubuntu server and upgraded it and used the local option instead, all good now.

Using: 
CSV 0.3.3
Grafana 7.4.3
